src/transfer.cpp

```diff
@@ -795,19 +817,51 @@ void Transfer::complete()
     else
     {
         LOG_debug << "Upload complete: " << (files.size() ? files.front()->name : "NO_FILES") << " " << files.size();
+        delete slot->fa;
+        slot->fa = NULL;
 
         // files must not change during a PUT transfer
-        if (slot->fa->asyncavailable())
+        for (file_list::iterator it = files.begin(); it != files.end(); )
         {
-            delete slot->fa;
-            slot->fa = client->fsaccess->newfileaccess();
-            if (!slot->fa->fopen(&localfilename))
+            File *f = (*it);
+            bool isOpen = true;
+            FileAccess *fa = client->fsaccess->newfileaccess();
+            if (!fa->fopen(&f->localname))
             {
-                return failed(API_EREAD);
+                isOpen = false;
+                if (client->fsaccess->transient_error)
+                {
+                    LOG_warn << "Retrying upload completion due to a transient error";
+                    slot->retrying = true;
+                    slot->retrybt.backoff(11);
+                    delete fa;
+                    return;
+                }
+            }
+
+            if (!isOpen || f->genfingerprint(fa))
+            {
+                LOG_warn << "Modification detected after upload";
+#ifdef ENABLE_SYNC
+                if (f->syncxfer)
+                {
+                    client->syncdownrequired = true;
+                }
+#endif
+                client->filecachedel(f);
+                files.erase(it++);
```

Comment:
The iterator can't be incremented before the erase, nor after because it would be invalid. That's why it's there. Another option would be to create a copy of the iterator (to be used inside the `if`), increment it (before the `if`). But I'm not sure if that approach is more clear than the current one.